### PR TITLE
Exported TE states to ThunderModule

### LINKED ISSUE #2438: Connect TEv2 states to ThunderModule
## 🚀 Feature

Make TEv2 states accessible as attributes to the thunder module. This would make it more transparent to the user and help with testing/verifying TEv2's accuracy.

### PR DESCRIPTION
**Pull request by mattteochen**
## What does this PR do?

Closes #2438.

### Summary
- Added ExportStatefulExecutorsTransform (singleton) with a registry of export callbacks.
- Executors can register a callback to export runtime state post-execution.
- Integrated for `TransformerEngine`: exports FP8 amax/scale and quantizer summaries to `module.te_fp8_stats`.

### Usage
The export is controlled by passing the transform to the transforms list:
```python
thunder.jit(model, executors=[...], transforms=[TransformerEngineTransform(), ExportStatefulExecutorsTransform()])
```
We might or might not go towards a compiler flag to make this more intuitive for the user (hence handling this transform automatically).

Whole example:
```python
from thunder.dev_utils.export_stateful_ex_transform import ExportStatefulExecutorsTransform
import torch
import torch.nn as nn
import thunder
from thunder.executors.transformer_engineex import transformer_engine_ex, TransformerEngineTransform
from transformer_engine.common import recipe
import transformer_engine.pytorch as te

torch.manual_seed(42)

# Device and data type
dtype = torch.bfloat16
device = "cuda"

# Inputs (3D input)
x = torch.randn(3, 768, 4096, device=device, dtype=dtype, requires_grad=True)

# TE recipe
fp8_recipe = recipe.DelayedScaling()

# Dummy model
class Module(nn.Module):
    def __init__(self):
        super().__init__()
        self.w1 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))
        self.w2 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))
        self.w3 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))

    def forward(self, x):
        return torch.nn.functional.linear(torch.nn.functional.linear(torch.nn.functional.linear(x, self.w2), self.w3), self.w1)

model = Module()
jmodel = thunder.jit(model, executors=[transformer_engine_ex], transforms=[TransformerEngineTransform(), ExportStatefulExecutorsTransform()])

# Enable autocasting for the forward pass
for _ in range(3):
    with te.fp8_autocast(fp8_recipe=fp8_recipe):
        y = jmodel(torch.randn(3, 768, 4096, device=device, dtype=dtype, requires_grad=True))
    grad_output = torch.randn_like(y)
    y.backward(grad_output)

# Visualize TE states
rep = jmodel.te_fp8_stats
from pprint import pprint
pprint(rep)
```

## Why it matters
- Not TE-specific: provides a reusable, uniform path for any stateful executor to export runtime state for debugging, validation, and reporting.

### CODE REVIEW

File: thunder/executors/transformer_engineex_impl.py [hunk 1]
```diff
@@ -35,15 +38,43 @@
...
+    if holder is not None:
+        tm_ref = getattr(holder, "_tm_ref", None)
+        if tm_ref is not None and isinstance(tm_ref, weakref.ref):
+            tm = tm_ref()
+
+    # Cannot fallback to compile data as during execution we don't have compile_data context
+    if tm is None:
+        import warnings
+
+        warnings.warn("No ThunderModule found for exporting TE states", UserWarning)
```
Comment (mattteochen) [inline thunder/executors/transformer_engineex_impl.py]: TE states are automatically exported; maybe we should consider adding a compilation or executor flag to let the user choose?

**Comment by mattteochen** [comment]
Test failure [logs](https://github.com/Lightning-AI/lightning-thunder/pull/2537/checks?check_run_id=51498425614) are not available somehow but should be related to #2444

File: thunder/dev_utils/te_states_reporter.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                "fp8_quant_bwd_grad": getattr(recipe, "fp8_quant_bwd_grad", None),
+                "fp8_gemm_fprop": getattr(recipe, "fp8_gemm_fprop", None),
+                "fp8_gemm_dgrad": getattr(recipe, "fp8_gemm_dgrad", None),
+                "fp8_gemm_wgrad": getattr(recipe, "fp8_gemm_wgrad", None),
+                "fp8_dpa": getattr(recipe, "fp8_dpa", None),
+                "fp8_mha": getattr(recipe, "fp8_mha", None),
+                "fwd_fp8_torch_dtype": str(get_fp8_torch_dtype(recipe, True)),
+                "bwd_fp8_torch_dtype": str(get_fp8_torch_dtype(recipe, False)),
+            }
+        )
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: TE executor does not support Float8CurrentScaling

File: thunder/dev_utils/te_states_reporter.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                "fp8_quant_fwd_inp": getattr(recipe, "fp8_quant_fwd_inp", None),
+                "fp8_quant_fwd_weight": getattr(recipe, "fp8_quant_fwd_weight", None),
+                "fp8_quant_bwd_grad": getattr(recipe, "fp8_quant_bwd_grad", None),
+                "fp8_gemm_fprop": getattr(recipe, "fp8_gemm_fprop", None),
+                "fp8_gemm_dgrad": getattr(recipe, "fp8_gemm_dgrad", None),
+                "fp8_gemm_wgrad": getattr(recipe, "fp8_gemm_wgrad", None),
+                "fwd_fp8_torch_dtype": str(get_fp8_torch_dtype(recipe, True)),
+                "bwd_fp8_torch_dtype": str(get_fp8_torch_dtype(recipe, False)),
+            }
+        )
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: TE executor does not support Float8BlockScaling

File: thunder/dev_utils/te_states_reporter.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        "columnwise_usage": getattr(quantizer, "columnwise_usage", None),
+        "internal": getattr(quantizer, "internal", None),
+        "dtype": str(getattr(quantizer, "dtype", None)),
+    }
+    # Optional attributes by quantizer class
+    if hasattr(quantizer, "with_amax_reduction"):
+        base["with_amax_reduction"] = getattr(quantizer, "with_amax_reduction")
+        base["amax_reduction_group"] = str(getattr(quantizer, "amax_reduction_group", None))
+    if hasattr(quantizer, "force_pow_2_scales"):
+        base["force_pow_2_scales"] = getattr(quantizer, "force_pow_2_scales")
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: Which recipe has a quantizer with this attribute?

File: thunder/dev_utils/te_states_reporter.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    recorded once per session for reporting correlation.
+
+    Returns:
+        A dictionary of global context fields suitable for rendering in reports.
+    """
+    fp8_enabled = FP8GlobalStateManager.is_fp8_enabled()
+    fp8_calibration = FP8GlobalStateManager.is_fp8_calibration()
+    with_fp8_params = FP8GlobalStateManager.with_fp8_parameters()
+    high_precision_init_val = FP8GlobalStateManager.with_high_precision_init_val()
+    fp8_group = FP8GlobalStateManager.get_fp8_group()
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: These are all internal TE states that are not used by the executor, why are they needed?

File: thunder/dev_utils/te_states_reporter.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    # Versions / device
+    cuda_version = getattr(torch.version, "cuda", None)
+    cublaslt_version = transformer_engine_torch.get_cublasLt_version()
+    device_cc = torch.cuda.get_device_capability() if torch.cuda.is_available() else None
+
+    # Dist info
+    world_size = None
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        try:
+            world_size = torch.distributed.get_world_size(group=fp8_group)
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: Is this the actual process group that is used in the executor?

https://github.com/Lightning-AI/lightning-thunder/blob/65193b0f2c3a710f3b7cf30a2b65cad18bb0622e/thunder/executors/transformer_engineex_impl.py#L330-L338

File: thunder/dev_utils/te_states_reporter.py [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+@dataclass
+class TEStateReporter:
+    """Accumulates TE runtime summaries and renders a report."""
+
+    global_ctx: dict[str, Any] | None = None
+    recipe_summaries: list[dict[str, Any]] = field(default_factory=list)
+    state_summaries_forward: list[dict[str, Any]] = field(default_factory=list)
+    state_summaries_backward: list[dict[str, Any]] = field(default_factory=list)
+    quantizer_summaries: list[dict[str, Any]] = field(default_factory=list)
+    shape_policy: dict[str, Any] = field(default_factory=lambda: {"mxfp8_block": MXFP8_BLOCK_SCALING_SIZE})
```
Comment (riccardofelluga) [inline thunder/dev_utils/te_states_reporter.py]: Might need to be updated after #2523

**Comment by mattteochen** [comment]
> This tool is a good idea! but let's limit the scope of this PR a little more. The issue was about connecting the thunder module to the states, however this PR introduces a great debugging tool to do similar thing.
> 
> Since this looks like to be a developer tool, let's limit it to be called only when in need. As per other dev tools, it would be better for this to accept a trace as input and work from there, without mixing itself deep down in the executor calls.
> 
> I'd say let's start simple and extract only amax and scale factors for the two supported recipes(ignore nvfp4 for now), what do you think?

Thank you for your suggestions @riccardofelluga!
I might be missing a point, though. If we make this a dev tool that accepts a trace and then performs actions, is exporting the states to a Thunder module still necessary?

Nesting a little within the executor calls was needed to recover the active main module to set these properties inside that entity. If we wanted to make this a standalone tool, this process would indeed no longer be necessary.

I see them as different solutions.
